src/util/make-knex.js

```diff
@@ -34,12 +34,21 @@ function initContext(knexFn) {
       return batchInsert(this, table, batch, chunkSize);
     },
 
-    // Runs a new transaction, taking a container and returning a promise
-    // for when the transaction is resolved.
+    // Creates a new transaction.
+    // If container is provided, returns a promise for when the transaction is resolved.
+    // If container is not provided, returns a promise with a transaction that is resolved
+    // when transaction is ready to be used.
     transaction(container, config) {
       const trx = this.client.transaction(container, config);
       trx.userParams = this.userParams;
-      return trx;
+
+      if (container) {
```

Comment:
So is `BEGIN TRANSACTION` query already sent to the DB server at this point?

Reply:
Isn't that what `this.client.transaction(container, config);` does?

Reply:
probably I was thinking originally that this PR would `Implement support for starting transactions without executing them` thus I could create transaction object, which would not allocate connection before the first query is ran (very useful in for example rest frameworks to provide transaction to handler which will not use any resources, if no queries are sent to there)... But I can see that this is just returning open and started transaction instead.

I was just confused about title of this PR :)

Reply:
right, this is that comment. this is totally feasible. do you think I should try implementing that?

Reply:
@elhigu Implemented transactionProvider for lazy instantiation of transactions that covers use-case that you've described. Feedback most welcome!